app/models/manageiq/providers/kubevirt/infra_manager/metrics_capture.rb

```
@@ -0,0 +1,139 @@
+
+    raise TargetValidationWarning, "no metrics endpoint found for #{target_name}" if metrics_connection(ems).nil?
+    raise TargetValidationWarning, "metrics authentication isn't valid for #{target_name}" unless metrics_connection_valid?(ems)
+  end
+
+  def build_capture_context!(ems, target, start_time, end_time)
+    verify_metrics_connection!(ems)
+    # make start_time align to minutes
+    start_time = start_time.beginning_of_minute
+
+    context = prometheus_capture_context(target, start_time, end_time)
+    raise TargetValidationWarning, "no metrics endpoint found for #{target_name}" if context.nil?
+
+    context
+  end
+
+  def perf_collect_metrics(interval_name, start_time = nil, end_time = nil)
+    start_time ||= 15.minutes.ago.beginning_of_minute.utc
+    ems = target.ext_management_system
+
+    _log.info("Collecting metrics for #{target_name} [#{interval_name}] [#{start_time}] [#{end_time}]")
+
+    begin
+      context = build_capture_context!(ems, target, start_time, end_time)
+    rescue TargetValidationError, TargetValidationWarning => e
+      _log.send(e.log_severity, "[#{target_name}] #{e.message}")
+      ems.try(:update,
+              :last_metrics_error       => :invalid,
+              :last_metrics_update_date => Time.now.utc)
+      raise
+    end
+
+    Benchmark.realtime_block(:collect_data) do
+      begin
+        context.collect_metrics
+      rescue NoMetricsFoundError => e
+        _log.warn("Metrics missing: [#{target_name}] #{e.message}")
+      rescue StandardError => e
+        _log.error("Metrics unavailable: [#{target_name}] #{e.message}")
+        ems.update(:last_metrics_error       => :unavailable,
+                              :last_metrics_update_date => Time.now.utc) if ems
+        raise
+      end
+    end
+
+    ems.update(:last_metrics_error        => nil,
+                          :last_metrics_update_date  => Time.now.utc,
+                          :last_metrics_success_date => Time.now.utc) if ems
```

Is it possible for ems to be nil here without something else failing early on?  Maybe an earlier check that target.ext_management_system is present?  (realize this is probably mostly copied but I can't unsee it)